pkg/util/crypto.go

```diff
@@ -0,0 +1,136 @@
+	klog.V(4).Infof(Log(ctx, "Encrypting device %s with LUKS"), devicePath)
+	if _, _, err := LuksFormat(devicePath, passphrase); err != nil {
+		return errors.Wrapf(err, "failed to encrypt device %s with LUKS", devicePath)
+	}
+	return nil
+}
+
+// OpenEncryptedVolume opens volume so that it can be used by the client
+func OpenEncryptedVolume(ctx context.Context, devicePath, mapperFile, passphrase string) error {
+	klog.V(4).Infof(Log(ctx, "Opening device %s with LUKS on %s"), devicePath, mapperFile)
+	_, _, err := LuksOpen(devicePath, mapperFile, passphrase)
+	return err
+}
+
+// CloseEncryptedVolume closes encrypted volume so it can be detached
+func CloseEncryptedVolume(ctx context.Context, mapperFile string) error {
+	klog.V(4).Infof(Log(ctx, "Closing LUKS device %s"), mapperFile)
+	_, _, err := LuksClose(mapperFile)
+	return err
+}
+
+// IsDeviceOpen determines if encrypted device is already open
+func IsDeviceOpen(device string) (bool, error) {
+	_, mappedFile, err := DeviceEncryptionStatus(device)
+	return (mappedFile != ""), err
+}
+
+// DeviceEncryptionStatus looks to identify if the passed device is a LUKS mapping
+// and if so what the device is and the mapper name as used by LUKS.
+// If not, just returns the original device and an empty string.
+func DeviceEncryptionStatus(devicePath string) (mappedDevice, mapper string, err error) {
+	if !strings.HasPrefix(devicePath, "/dev/mapper") {
+		return devicePath, "", nil
+	}
+	mapPath := strings.TrimPrefix(devicePath, "/dev/mapper/")
+	stdout, _, err := LuksStatus(mapPath)
+	if err != nil {
+		// Non-zero return implies not an active LUKS device
+		return devicePath, "", nil
+	}
+	lines := strings.Split(string(stdout), "\n")
+	if !strings.HasSuffix(lines[0], " is active.") {
+		// Implies this is not a LUKS device
+		return devicePath, "", nil
+	}
+	for i := 1; i < len(lines); i++ {
+		kv := strings.SplitN(strings.TrimSpace(lines[i]), ":", 2)
+		if strings.Compare(kv[0], "device") == 0 {
+			return strings.TrimSpace(kv[1]), mapPath, nil
+		}
+	}
+	// Identified as LUKS, but failed to identify a mapped device
+	return "", "", errors.New("mapped device not found")
+}
+
+// CheckRbdImageEncrypted verifies if rbd image was encrypted when created
+func CheckRbdImageEncrypted(ctx context.Context, cr *Credentials, monitors, imageSpec string) (string, error) {
+	value, err := GetImageMeta(ctx, cr, monitors, imageSpec, encryptionMetaKey)
+	if err != nil {
+		klog.Errorf(Log(ctx, "checking image %s encrypted state metadata failed: %s"), imageSpec, err)
+		return "", err
+	}
+
+	encrypted := strings.TrimSpace(value)
+	klog.V(4).Infof(Log(ctx, "image %s encrypted state metadata reports '%s'"), imageSpec, encrypted)
+	return encrypted, nil
+}
+
+// SaveRbdImageEncryptionStatus sets image metadata for encryption status
+func SaveRbdImageEncryptionStatus(ctx context.Context, cr *Credentials, monitors, imageSpec, status string) error {
```

Comment:
I was thinking can  we just save the image metadata with rbd image itself, not with rados omap

example:
```
sh-4.2# rbd image-meta set replicapool/test encrypted true
sh-4.2# rbd image-meta list replicapool/test
There is 1 metadatum on this image:

Key       Value 
encrypted true 
sh-4.2# rbd image-meta get replicapool/test encrypted
true
sh-4.2# rbd image-meta remove replicapool/test encrypted
```

Reply:
@dillaman @ShyamsundarR ^^

Reply:
that's what SetImageMeta is doing, it is using `rbd image-meta set`

Reply:
sorry I missed  it

Reply:
as discussed offline: the confusion came from testing part where I was looking for rados omap value to map pvc to VolumeID in ceph, but as it appears there is a way to do so and I've changed that now.